<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>calour.experiment.Experiment &#8212; calour 2018.10.1 documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="calour.experiment.Experiment.__deepcopy__" href="calour.experiment.Experiment.__deepcopy__.html" />
    <link rel="prev" title="experiment (calour.experiment)" href="../experiment.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head><body>
  
  <a href="https://github.com/biocore/calour"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          calour docs</a>
        <span class="navbar-text navbar-version pull-left"><b>2018.10.1</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">General Calour concepts</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials_microbiome.html">Microbiome tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials_metabolome.html">Metabolome tutorials</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../api.html">Using Calour functions via the module and the Experiment class</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html#main-classes-and-utility-functions">Main classes and utility functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#functions-operating-on-experiment-object">Functions operating on <code class="docutils literal notranslate"><span class="pre">Experiment</span></code> object</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../database_interface.html">Creating database interfaces for Calour</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">calour.experiment.Experiment</a><ul>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="calour-experiment-experiment">
<h1>calour.experiment.Experiment<a class="headerlink" href="#calour-experiment-experiment" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="calour.experiment.Experiment">
<em class="property">class </em><code class="descclassname">calour.experiment.</code><code class="descname">Experiment</code><span class="sig-paren">(</span><em>data</em>, <em>sample_metadata</em>, <em>feature_metadata=None</em>, <em>exp_metadata={}</em>, <em>description=''</em>, <em>sparse=True</em><span class="sig-paren">)</span><a class="reference external" href="http://github.com/biocore/calour/blob/2018.10.1/calour/experiment.py#L36"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#calour.experiment.Experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/dev/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>This class contains the data for a experiment or a meta experiment.</p>
<p>The data set includes a data table (otu table, gene table,
metabolomic table, or all those tables combined), a sample
metadata table, and a feature metadata.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.15)"><em>numpy.ndarray</em></a><em> or </em><a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix" title="(in SciPy v1.1.0)"><em>scipy.sparse.csr_matrix</em></a>) – The abundance table for OTUs, metabolites, genes, etc. Samples
are in row and features in column</li>
<li><strong>sample_metadata</strong> (<a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v0.23.4)"><em>pandas.DataFrame</em></a>) – The metadata on the samples</li>
<li><strong>feature_metadata</strong> (<a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v0.23.4)"><em>pandas.DataFrame</em></a>) – The metadata on the features</li>
<li><strong>description</strong> (<a class="reference external" href="https://docs.python.org/dev/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – name of experiment</li>
<li><strong>sparse</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – store the data array in <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix" title="(in SciPy v1.1.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">scipy.sparse.csr_matrix</span></code></a>
or <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.15)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a></li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first last simple">
<li><strong>data</strong> (<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.15)"><em>numpy.ndarray</em></a><em> or </em><a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix" title="(in SciPy v1.1.0)"><em>scipy.sparse.csr_matrix</em></a>) – The abundance table for OTUs, metabolites, genes, etc. Samples
are in row and features in column</li>
<li><strong>sample_metadata</strong> (<a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v0.23.4)"><em>pandas.DataFrame</em></a>) – The metadata on the samples</li>
<li><strong>feature_metadata</strong> (<a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v0.23.4)"><em>pandas.DataFrame</em></a>) – The metadata on the features</li>
<li><strong>exp_metadata</strong> (<a class="reference external" href="https://docs.python.org/dev/library/stdtypes.html#dict" title="(in Python v3.8)"><em>dict</em></a>) – metadata about the experiment (data md5, filenames, etc.)</li>
<li><a class="reference internal" href="calour.amplicon_experiment.AmpliconExperiment.shape.html#calour.amplicon_experiment.AmpliconExperiment.shape" title="calour.amplicon_experiment.AmpliconExperiment.shape"><strong>shape</strong></a> (<em>tuple of</em><em> (</em><a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>)</em>) – the dimension of data</li>
<li><a class="reference internal" href="calour.amplicon_experiment.AmpliconExperiment.sparse.html#calour.amplicon_experiment.AmpliconExperiment.sparse" title="calour.amplicon_experiment.AmpliconExperiment.sparse"><strong>sparse</strong></a> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – store the data array in <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix" title="(in SciPy v1.1.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">scipy.sparse.csr_matrix</span></code></a>
or <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.15)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a></li>
<li><strong>normalized</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – the normalization factor. it is zero if not normalized</li>
<li><strong>description</strong> (<a class="reference external" href="https://docs.python.org/dev/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – name of the experiment</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><code class="xref py py-class docutils literal notranslate"><span class="pre">AmpliconExperiment</span></code></p>
</div>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.__deepcopy__.html#calour.experiment.Experiment.__deepcopy__" title="calour.experiment.Experiment.__deepcopy__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy.deepcopy(exp)</span></code></a></td>
<td>Implement the deepcopy since pandas has problem deepcopy empty dataframe</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.__eq__.html#calour.experiment.Experiment.__eq__" title="calour.experiment.Experiment.__eq__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">exp1</span> <span class="pre">==</span> <span class="pre">exp2</span></code></a></td>
<td>Check equality.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.__getitem__.html#calour.experiment.Experiment.__getitem__" title="calour.experiment.Experiment.__getitem__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">exp[k]</span></code></a></td>
<td>Get the abundance at (sampleid, featureid)</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.__ne__.html#calour.experiment.Experiment.__ne__" title="calour.experiment.Experiment.__ne__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">exp1</span> <span class="pre">!=</span> <span class="pre">exp2</span></code></a></td>
<td>Return self!=value.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.__repr__.html#calour.experiment.Experiment.__repr__" title="calour.experiment.Experiment.__repr__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">repr(exp)</span></code></a></td>
<td>Return a string representation of this object.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.add_sample_metadata_as_features.html#calour.experiment.Experiment.add_sample_metadata_as_features" title="calour.experiment.Experiment.add_sample_metadata_as_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_sample_metadata_as_features</span></code></a>(fields[,&nbsp;…])</td>
<td>Add covariates from sample metadata to the data table as features for machine learning.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.add_terms_to_features.html#calour.experiment.Experiment.add_terms_to_features" title="calour.experiment.Experiment.add_terms_to_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_terms_to_features</span></code></a>(dbname[,&nbsp;…])</td>
<td>Add a field to the feature metadata, with most common term for each feature</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.aggregate_by_metadata.html#calour.experiment.Experiment.aggregate_by_metadata" title="calour.experiment.Experiment.aggregate_by_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">aggregate_by_metadata</span></code></a>(field[,&nbsp;agg,&nbsp;axis,&nbsp;…])</td>
<td>aggregate all samples/features that have the same value in the given field.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.binarize.html#calour.experiment.Experiment.binarize" title="calour.experiment.Experiment.binarize"><code class="xref py py-obj docutils literal notranslate"><span class="pre">binarize</span></code></a>([threshold,&nbsp;inplace])</td>
<td>Binarize the data with a threshold.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.center_log_ratio.html#calour.experiment.Experiment.center_log_ratio" title="calour.experiment.Experiment.center_log_ratio"><code class="xref py py-obj docutils literal notranslate"><span class="pre">center_log_ratio</span></code></a>([method,&nbsp;centralize,&nbsp;inplace])</td>
<td>Performs a clr transform to normalize each sample.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.classify.html#calour.experiment.Experiment.classify" title="calour.experiment.Experiment.classify"><code class="xref py py-obj docutils literal notranslate"><span class="pre">classify</span></code></a>(fields,&nbsp;estimator[,&nbsp;cv,&nbsp;predict,&nbsp;…])</td>
<td>Evaluate classification during cross validation.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.cluster_data.html#calour.experiment.Experiment.cluster_data" title="calour.experiment.Experiment.cluster_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cluster_data</span></code></a>([transform,&nbsp;axis,&nbsp;metric,&nbsp;inplace])</td>
<td>Cluster the samples/features.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.cluster_features.html#calour.experiment.Experiment.cluster_features" title="calour.experiment.Experiment.cluster_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cluster_features</span></code></a>([min_abundance,&nbsp;inplace])</td>
<td>Cluster features.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.copy.html#calour.experiment.Experiment.copy" title="calour.experiment.Experiment.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</td>
<td>Copy the object (deeply).</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.correlation.html#calour.experiment.Experiment.correlation" title="calour.experiment.Experiment.correlation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">correlation</span></code></a>(field[,&nbsp;method,&nbsp;nonzero,&nbsp;…])</td>
<td>Find features with correlation to a numeric metadata field</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.diff_abundance.html#calour.experiment.Experiment.diff_abundance" title="calour.experiment.Experiment.diff_abundance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">diff_abundance</span></code></a>(field,&nbsp;val1[,&nbsp;val2,&nbsp;method,&nbsp;…])</td>
<td>test the differential expression between 2 groups (val1 and val2 in field field) using permutation based fdr (dsfdr) for bacteria that have a significant difference.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.diff_abundance_kw.html#calour.experiment.Experiment.diff_abundance_kw" title="calour.experiment.Experiment.diff_abundance_kw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">diff_abundance_kw</span></code></a>(field[,&nbsp;transform,&nbsp;…])</td>
<td>Test the differential expression between multiple sample groups using the Kruskal Wallis test.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.downsample.html#calour.experiment.Experiment.downsample" title="calour.experiment.Experiment.downsample"><code class="xref py py-obj docutils literal notranslate"><span class="pre">downsample</span></code></a>(field[,&nbsp;axis,&nbsp;num_keep,&nbsp;inplace,&nbsp;…])</td>
<td>Downsample the data set.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.enrichment.html#calour.experiment.Experiment.enrichment" title="calour.experiment.Experiment.enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">enrichment</span></code></a>(features,&nbsp;dbname,&nbsp;*args,&nbsp;**kwargs)</td>
<td>Get the list of enriched annotation terms in features compared to all other features in exp.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.export_html.html#calour.experiment.Experiment.export_html" title="calour.experiment.Experiment.export_html"><code class="xref py py-obj docutils literal notranslate"><span class="pre">export_html</span></code></a>([sample_field,&nbsp;feature_field,&nbsp;…])</td>
<td>Export an interactive html heatmap for the experiment.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.filter_abundance.html#calour.experiment.Experiment.filter_abundance" title="calour.experiment.Experiment.filter_abundance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_abundance</span></code></a>([cutoff])</td>
<td>Filter features with sum abundance across all samples less than the cutoff.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.filter_by_data.html#calour.experiment.Experiment.filter_by_data" title="calour.experiment.Experiment.filter_by_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_by_data</span></code></a>(predicate[,&nbsp;axis,&nbsp;field,&nbsp;…])</td>
<td>Filter samples or features by the data matrix.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.filter_by_metadata.html#calour.experiment.Experiment.filter_by_metadata" title="calour.experiment.Experiment.filter_by_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_by_metadata</span></code></a>(field,&nbsp;select[,&nbsp;axis,&nbsp;…])</td>
<td>Filter samples or features by metadata.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.filter_ids.html#calour.experiment.Experiment.filter_ids" title="calour.experiment.Experiment.filter_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_ids</span></code></a>(ids[,&nbsp;axis,&nbsp;negate,&nbsp;inplace])</td>
<td>Filter samples or features based on a list IDs.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.filter_mean_abundance.html#calour.experiment.Experiment.filter_mean_abundance" title="calour.experiment.Experiment.filter_mean_abundance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_mean_abundance</span></code></a>([cutoff,&nbsp;field])</td>
<td>Filter features with a mean at least cutoff of the mean total abundance/sample</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.filter_prevalence.html#calour.experiment.Experiment.filter_prevalence" title="calour.experiment.Experiment.filter_prevalence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_prevalence</span></code></a>(fraction[,&nbsp;cutoff,&nbsp;field])</td>
<td>Filter features keeping only ones present in more than certain fraction of all samples.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.filter_sample_categories.html#calour.experiment.Experiment.filter_sample_categories" title="calour.experiment.Experiment.filter_sample_categories"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_sample_categories</span></code></a>(field[,&nbsp;…])</td>
<td>Filter sample categories that have too few samples.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.filter_samples.html#calour.experiment.Experiment.filter_samples" title="calour.experiment.Experiment.filter_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_samples</span></code></a>(field,&nbsp;values[,&nbsp;negate,&nbsp;inplace])</td>
<td>Shortcut for filtering samples.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.filter_samples_.html#calour.experiment.Experiment.filter_samples_" title="calour.experiment.Experiment.filter_samples_"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_samples_</span></code></a>([cutoff,&nbsp;inplace])</td>
<td></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.from_pandas.html#calour.experiment.Experiment.from_pandas" title="calour.experiment.Experiment.from_pandas"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_pandas</span></code></a>(df[,&nbsp;exp])</td>
<td>Convert a Pandas DataFrame into an experiment.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.get_data.html#calour.experiment.Experiment.get_data" title="calour.experiment.Experiment.get_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_data</span></code></a>([sparse,&nbsp;copy])</td>
<td>Get the data as a 2d array</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.heatmap.html#calour.experiment.Experiment.heatmap" title="calour.experiment.Experiment.heatmap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">heatmap</span></code></a>([sample_field,&nbsp;feature_field,&nbsp;…])</td>
<td>Plot a heatmap for the experiment.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.join_experiments.html#calour.experiment.Experiment.join_experiments" title="calour.experiment.Experiment.join_experiments"><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_experiments</span></code></a>(other[,&nbsp;field_name,&nbsp;prefixes])</td>
<td>Combine two <a class="reference internal" href="#calour.experiment.Experiment" title="calour.experiment.Experiment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Experiment</span></code></a> objects into one.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.join_experiments_featurewise.html#calour.experiment.Experiment.join_experiments_featurewise" title="calour.experiment.Experiment.join_experiments_featurewise"><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_experiments_featurewise</span></code></a>(other[,&nbsp;…])</td>
<td>Combine two <a class="reference internal" href="#calour.experiment.Experiment" title="calour.experiment.Experiment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Experiment</span></code></a> objects into one.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.join_metadata_fields.html#calour.experiment.Experiment.join_metadata_fields" title="calour.experiment.Experiment.join_metadata_fields"><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_metadata_fields</span></code></a>(field1,&nbsp;field2[,&nbsp;…])</td>
<td>Join two sample/feature metadata fields into a single new field</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.log_n.html#calour.experiment.Experiment.log_n" title="calour.experiment.Experiment.log_n"><code class="xref py py-obj docutils literal notranslate"><span class="pre">log_n</span></code></a>([n,&nbsp;inplace])</td>
<td>Log transform the data</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.normalize.html#calour.experiment.Experiment.normalize" title="calour.experiment.Experiment.normalize"><code class="xref py py-obj docutils literal notranslate"><span class="pre">normalize</span></code></a>([total,&nbsp;axis,&nbsp;inplace])</td>
<td>Normalize the sum of each sample (axis=0) or feature (axis=1) to sum total</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.normalize_by_subset_features.html#calour.experiment.Experiment.normalize_by_subset_features" title="calour.experiment.Experiment.normalize_by_subset_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">normalize_by_subset_features</span></code></a>(features[,&nbsp;…])</td>
<td>Normalize each sample by their total sums without a list of features</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.normalize_compositional.html#calour.experiment.Experiment.normalize_compositional" title="calour.experiment.Experiment.normalize_compositional"><code class="xref py py-obj docutils literal notranslate"><span class="pre">normalize_compositional</span></code></a>([min_frac,&nbsp;total,&nbsp;…])</td>
<td>Normalize each sample by ignoring the features with mean&gt;=min_frac in all the experiment</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.plot.html#calour.experiment.Experiment.plot" title="calour.experiment.Experiment.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>([title,&nbsp;barx_fields,&nbsp;barx_width,&nbsp;…])</td>
<td>Plot the interactive heatmap and its associated axes.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.plot_abund_prevalence.html#calour.experiment.Experiment.plot_abund_prevalence" title="calour.experiment.Experiment.plot_abund_prevalence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_abund_prevalence</span></code></a>(field[,&nbsp;log,&nbsp;…])</td>
<td>Plot abundance against prevalence.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.plot_core_features.html#calour.experiment.Experiment.plot_core_features" title="calour.experiment.Experiment.plot_core_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_core_features</span></code></a>([field,&nbsp;steps,&nbsp;cutoff,&nbsp;…])</td>
<td>Plot the percentage of core features shared in increasing number of samples.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.plot_diff_abundance_enrichment.html#calour.experiment.Experiment.plot_diff_abundance_enrichment" title="calour.experiment.Experiment.plot_diff_abundance_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_diff_abundance_enrichment</span></code></a>([max_show,&nbsp;…])</td>
<td>Plot the term enrichment of differentially abundant bacteria</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.plot_enrichment.html#calour.experiment.Experiment.plot_enrichment" title="calour.experiment.Experiment.plot_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_enrichment</span></code></a>(enriched[,&nbsp;max_show,&nbsp;…])</td>
<td>Plot a horizontal bar plot for enriched terms</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.plot_feature_matrix.html#calour.experiment.Experiment.plot_feature_matrix" title="calour.experiment.Experiment.plot_feature_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_feature_matrix</span></code></a>(fields,&nbsp;feature_ids[,&nbsp;…])</td>
<td>This plots an array of scatter plots between each features against the specified sample metadata.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.plot_hist.html#calour.experiment.Experiment.plot_hist" title="calour.experiment.Experiment.plot_hist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_hist</span></code></a>([ax])</td>
<td>Plot histogram of all the values in data.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.plot_stacked_bar.html#calour.experiment.Experiment.plot_stacked_bar" title="calour.experiment.Experiment.plot_stacked_bar"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_stacked_bar</span></code></a>([field,&nbsp;sample_color_bars,&nbsp;…])</td>
<td>Plot the stacked bar for feature abundances.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.random_permute_data.html#calour.experiment.Experiment.random_permute_data" title="calour.experiment.Experiment.random_permute_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">random_permute_data</span></code></a>([normalize])</td>
<td>Shuffle independently the reads of each feature</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.regress.html#calour.experiment.Experiment.regress" title="calour.experiment.Experiment.regress"><code class="xref py py-obj docutils literal notranslate"><span class="pre">regress</span></code></a>(field,&nbsp;estimator[,&nbsp;cv,&nbsp;params])</td>
<td>Evaluate regression during cross validation.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.reorder.html#calour.experiment.Experiment.reorder" title="calour.experiment.Experiment.reorder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reorder</span></code></a>(new_order[,&nbsp;axis,&nbsp;inplace])</td>
<td>Reorder according to indices in the new order.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.rescale.html#calour.experiment.Experiment.rescale" title="calour.experiment.Experiment.rescale"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rescale</span></code></a>([total,&nbsp;axis,&nbsp;inplace])</td>
<td>Rescale the data to mean sum of all samples (axis=0) or features (axis=1) to be total.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.save.html#calour.experiment.Experiment.save" title="calour.experiment.Experiment.save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save</span></code></a>(prefix[,&nbsp;fmt])</td>
<td>Save the experiment data to disk.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.save_biom.html#calour.experiment.Experiment.save_biom" title="calour.experiment.Experiment.save_biom"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_biom</span></code></a>(f[,&nbsp;fmt,&nbsp;add_metadata])</td>
<td>Save experiment to biom format</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.save_fasta.html#calour.experiment.Experiment.save_fasta" title="calour.experiment.Experiment.save_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code></a>(f[,&nbsp;seqs])</td>
<td>Save a list of sequences to fasta.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.save_metadata.html#calour.experiment.Experiment.save_metadata" title="calour.experiment.Experiment.save_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_metadata</span></code></a>(f[,&nbsp;axis])</td>
<td>Save sample/feature metadata to file.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.scale.html#calour.experiment.Experiment.scale" title="calour.experiment.Experiment.scale"><code class="xref py py-obj docutils literal notranslate"><span class="pre">scale</span></code></a>([axis,&nbsp;inplace])</td>
<td>Standardize a dataset along an axis</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.sort_abundance.html#calour.experiment.Experiment.sort_abundance" title="calour.experiment.Experiment.sort_abundance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort_abundance</span></code></a>([subgroup])</td>
<td>Sort features based on their abundance in a subset of the samples.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.sort_by_data.html#calour.experiment.Experiment.sort_by_data" title="calour.experiment.Experiment.sort_by_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort_by_data</span></code></a>([axis,&nbsp;subset,&nbsp;key,&nbsp;inplace,&nbsp;…])</td>
<td>Sort features based on their mean frequency.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.sort_by_metadata.html#calour.experiment.Experiment.sort_by_metadata" title="calour.experiment.Experiment.sort_by_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort_by_metadata</span></code></a>(field[,&nbsp;axis,&nbsp;inplace])</td>
<td>Sort samples or features based on metadata values in the field.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.sort_centroid.html#calour.experiment.Experiment.sort_centroid" title="calour.experiment.Experiment.sort_centroid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort_centroid</span></code></a>([transform,&nbsp;inplace])</td>
<td>Sort the features based on the center of mass</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.sort_ids.html#calour.experiment.Experiment.sort_ids" title="calour.experiment.Experiment.sort_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort_ids</span></code></a>(ids[,&nbsp;axis,&nbsp;inplace])</td>
<td>Sort the features or samples by the given ids.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.sort_samples.html#calour.experiment.Experiment.sort_samples" title="calour.experiment.Experiment.sort_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort_samples</span></code></a>(field,&nbsp;**kwargs)</td>
<td>Sort samples by field A convenience function for sort_by_metadata</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.split_train_test.html#calour.experiment.Experiment.split_train_test" title="calour.experiment.Experiment.split_train_test"><code class="xref py py-obj docutils literal notranslate"><span class="pre">split_train_test</span></code></a>(test_size[,&nbsp;train_size,&nbsp;…])</td>
<td>Split experiment into train experiment and test experiment.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.subsample_count.html#calour.experiment.Experiment.subsample_count" title="calour.experiment.Experiment.subsample_count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subsample_count</span></code></a>(total[,&nbsp;replace,&nbsp;inplace,&nbsp;…])</td>
<td>Randomly subsample each sample to the same number of counts.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.to_pandas.html#calour.experiment.Experiment.to_pandas" title="calour.experiment.Experiment.to_pandas"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_pandas</span></code></a>([sample_field,&nbsp;feature_field,&nbsp;sparse])</td>
<td>Get a pandas dataframe of the abundances Samples are rows, features are columns.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.transform.html#calour.experiment.Experiment.transform" title="calour.experiment.Experiment.transform"><code class="xref py py-obj docutils literal notranslate"><span class="pre">transform</span></code></a>([steps,&nbsp;inplace])</td>
<td>Chain transformations together.</td>
</tr>
</tbody>
</table>
<p class="rubric">Attributes</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="calour.experiment.Experiment.shape.html#calour.experiment.Experiment.shape" title="calour.experiment.Experiment.shape"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shape</span></code></a></td>
<td></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="calour.experiment.Experiment.sparse.html#calour.experiment.Experiment.sparse" title="calour.experiment.Experiment.sparse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sparse</span></code></a></td>
<td></td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2016-2018, calour development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.7.3.<br/>
    </p>
  </div>
</footer>
  </body>
</html>