<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
        <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Micribiome data manipulation tutorial &#8212; calour 1.0-dev documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Micribiome differential abundance tutorial" href="microbiome_diff_abundance.html" />
    <link rel="prev" title="Micribiome experiment step-by-step analysis" href="microbiome_step_by_step.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head><body>
  
  <a href="https://github.com/biocore/calour"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          calour docs</a>
        <span class="navbar-text navbar-version pull-left"><b>1.0-dev</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">General Calour concepts</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials_microbiome.html">Microbiome tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials_metabolome.html">Metabolome tutorials</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">Using Calour functions via the module and the Experiment class</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#main-classes-and-utility-functions">Main classes and utility functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#functions-operating-on-experiment-object">Functions operating on <code class="docutils literal notranslate"><span class="pre">Experiment</span></code> object</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../database_interface.html">Creating database interfaces for Calour</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Micribiome data manipulation tutorial</a><ul>
<li><a class="reference internal" href="#Setup">Setup</a></li>
<li><a class="reference internal" href="#Load-the-data">Load the data</a></li>
<li><a class="reference internal" href="#sorting-the-samples-based-on-a-metadata-field-(sort_samples)">sorting the samples based on a metadata field (<code class="docutils literal notranslate"><span class="pre">sort_samples</span></code>)</a><ul>
<li><a class="reference internal" href="#consecutive-sorting-using-different-fields">consecutive sorting using different fields</a></li>
</ul>
</li>
<li><a class="reference internal" href="#filter-samples-based-on-metadata-field-(filter_samples)">filter samples based on metadata field (<code class="docutils literal notranslate"><span class="pre">filter_samples</span></code>)</a><ul>
<li><a class="reference internal" href="#we-can-supply-a-list-of-values-instead-of-only-one-value">we can supply a list of values instead of only one value</a></li>
<li><a class="reference internal" href="#we-can-also-reverse-the-filtering-(removing-samples-with-the-supplied-values)">we can also reverse the filtering (removing samples with the supplied values)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#filter-low-abundance-features-(filter_abundance)">filter low abundance features (<code class="docutils literal notranslate"><span class="pre">filter_abundance</span></code>)</a><ul>
<li><a class="reference internal" href="#Keeping-the-low-abundance-bacteria-instead">Keeping the low abundance bacteria instead</a></li>
</ul>
</li>
<li><a class="reference internal" href="#filter-non-common-bacteria-(filter_prevalence)">filter non-common bacteria (<code class="docutils literal notranslate"><span class="pre">filter_prevalence</span></code>)</a></li>
<li><a class="reference internal" href="#Filter-bacteria-based-on-the-mean-frequency-over-all-samples-(filter_mean)">Filter bacteria based on the mean frequency over all samples (<code class="docutils literal notranslate"><span class="pre">filter_mean</span></code>)</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput,
div.nbinput div.prompt,
div.nbinput div.input_area,
div.nbinput div[class*=highlight],
div.nbinput div[class*=highlight] pre,
div.nboutput,
div.nbinput div.prompt,
div.nbinput div.output_area,
div.nboutput div[class*=highlight],
div.nboutput div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput,
div.nboutput {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput,
    div.nboutput {
        flex-direction: column;
    }
}

/* input container */
div.nbinput {
    padding-top: 5px;
}

/* last container */
div.nblast {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput div.prompt pre {
    color: #303F9F;
}

/* output prompt */
div.nboutput div.prompt pre {
    color: #D84315;
}

/* all prompts */
div.nbinput div.prompt,
div.nboutput div.prompt {
    min-width: 9ex;
    padding-top: 0.4em;
    padding-right: 0.4em;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput div.prompt,
    div.nboutput div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput div.prompt pre,
div.nboutput div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput div.input_area,
div.nboutput div.output_area {
    padding: 0.4em;
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput div.input_area,
    div.nboutput div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput div.input_area {
    border: 1px solid #cfcfcf;
    border-radius: 2px;
    background: #f7f7f7;
}

/* override MathJax center alignment in output cells */
div.nboutput div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.pngmath center alignment in output cells */
div.nboutput div.math p {
    text-align: left;
}

/* standard error */
div.nboutput div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }
</style>
<div class="section" id="Micribiome-data-manipulation-tutorial">
<h1>Micribiome data manipulation tutorial<a class="headerlink" href="#Micribiome-data-manipulation-tutorial" title="Permalink to this headline">¶</a></h1>
<p>This is a jupyter notebook example of how to sort, filter and handle
sample metadata</p>
<div class="section" id="Setup">
<h2>Setup<a class="headerlink" href="#Setup" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">calour</span> <span class="kn">as</span> <span class="nn">ca</span>
<span class="n">ca</span><span class="o">.</span><span class="n">set_log_level</span><span class="p">(</span><span class="mi">11</span><span class="p">)</span>
<span class="o">%</span><span class="k">matplotlib</span> notebook
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="stderr output_area docutils container">
<div class="highlight"><pre>
/Users/amnon/miniconda3/envs/calour/lib/python3.5/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
</pre></div></div>
</div>
</div>
<div class="section" id="Load-the-data">
<h2>Load the data<a class="headerlink" href="#Load-the-data" title="Permalink to this headline">¶</a></h2>
<p>we use two datasets:</p>
<p>the Chronic faitigue syndrome data from:</p>
<p>Giloteaux, L., Goodrich, J.K., Walters, W.A., Levine, S.M., Ley, R.E.
and Hanson, M.R., 2016.</p>
<p>Reduced diversity and altered composition of the gut microbiome in
individuals with myalgic encephalomyelitis/chronic fatigue syndrome.</p>
<p>Microbiome, 4(1), p.30.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cfs</span><span class="o">=</span><span class="n">ca</span><span class="o">.</span><span class="n">read_amplicon</span><span class="p">(</span><span class="s1">&#39;data/chronic-fatigue-syndrome.biom&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;data/chronic-fatigue-syndrome.sample.txt&#39;</span><span class="p">,</span>
                     <span class="n">normalize</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span><span class="n">min_reads</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2018-03-04 12:38:34 INFO loaded 87 samples, 2129 features
2018-03-04 12:38:34 WARNING These have metadata but do not have data - dropped: {&#39;ERR1331814&#39;}
2018-03-04 12:38:34 INFO After filtering, 87 remaining
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">print</span><span class="p">(</span><span class="n">cfs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AmpliconExperiment (&#34;chronic-fatigue-syndrome.biom&#34;) with 87 samples, 2129 features
</pre></div></div>
</div>
<p>Moving pictures dataset. from:</p>
<p>Caporaso, J.G., Lauber, C.L., Costello, E.K., Berg-Lyons, D., Gonzalez,
A., Stombaugh, J., Knights, D., Gajer, P., Ravel, J., Fierer, N. and
Gordon, J.I., 2011.</p>
<p>Moving pictures of the human microbiome.</p>
<p>Genome biology, 12(5), p.R50.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">movpic</span><span class="o">=</span><span class="n">ca</span><span class="o">.</span><span class="n">read_amplicon</span><span class="p">(</span><span class="s1">&#39;data/moving_pic.biom&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;data/moving_pic.sample.txt&#39;</span><span class="p">,</span>
                     <span class="n">normalize</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span><span class="n">min_reads</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2018-03-04 12:38:36 INFO loaded 1968 samples, 7056 features
2018-03-04 12:38:37 INFO After filtering, 1967 remaining
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">print</span><span class="p">(</span><span class="n">movpic</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AmpliconExperiment (&#34;moving_pic.biom&#34;) with 1967 samples, 7056 features
</pre></div></div>
</div>
</div>
<div class="section" id="sorting-the-samples-based-on-a-metadata-field-(sort_samples)">
<h2>sorting the samples based on a metadata field (<code class="docutils literal notranslate"><span class="pre">sort_samples</span></code>)<a class="headerlink" href="#sorting-the-samples-based-on-a-metadata-field-(sort_samples)" title="Permalink to this headline">¶</a></h2>
<p>Sort the samples of the experiment based on the values in the given
field.</p>
<p>is the original data sorted by the Subject field?</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">print</span><span class="p">(</span><span class="n">cfs</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">[</span><span class="s1">&#39;Subject&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">is_monotonic_increasing</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
False
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cfs</span><span class="o">=</span><span class="n">cfs</span><span class="o">.</span><span class="n">sort_samples</span><span class="p">(</span><span class="s1">&#39;Subject&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>and is the new data sorted?</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">print</span><span class="p">(</span><span class="n">cfs</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">[</span><span class="s1">&#39;Subject&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">is_monotonic_increasing</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
True
</pre></div></div>
</div>
<div class="section" id="consecutive-sorting-using-different-fields">
<h3>consecutive sorting using different fields<a class="headerlink" href="#consecutive-sorting-using-different-fields" title="Permalink to this headline">¶</a></h3>
<p>Keeps the order of the previous fields if values for the new field are
tied.</p>
<p>For the moving pictures dataset, we want the data to be sorted by
individual, and within each individual to be sorted by timepoint</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">movpic</span><span class="o">=</span><span class="n">movpic</span><span class="o">.</span><span class="n">sort_samples</span><span class="p">(</span><span class="s1">&#39;DAYS_SINCE_EXPERIMENT_START&#39;</span><span class="p">)</span>
<span class="n">movpic</span><span class="o">=</span><span class="n">movpic</span><span class="o">.</span><span class="n">sort_samples</span><span class="p">(</span><span class="s1">&#39;HOST_SUBJECT_ID&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">print</span><span class="p">(</span><span class="n">movpic</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">[</span><span class="s1">&#39;DAYS_SINCE_EXPERIMENT_START&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">is_monotonic_increasing</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
False
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">print</span><span class="p">(</span><span class="n">movpic</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">[</span><span class="s1">&#39;HOST_SUBJECT_ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">is_monotonic_increasing</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
True
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="filter-samples-based-on-metadata-field-(filter_samples)">
<h2>filter samples based on metadata field (<code class="docutils literal notranslate"><span class="pre">filter_samples</span></code>)<a class="headerlink" href="#filter-samples-based-on-metadata-field-(filter_samples)" title="Permalink to this headline">¶</a></h2>
<p>Keep only samples matching the values we supply for the selected
metadata field.</p>
<p>lets keep only samples from participant F4</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">tt</span><span class="o">=</span><span class="n">movpic</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="s1">&#39;HOST_SUBJECT_ID&#39;</span><span class="p">,</span><span class="s1">&#39;F4&#39;</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s1">&#39;* original:</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n\n</span><span class="s1">* filtered:</span><span class="se">\n</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">movpic</span><span class="p">,</span> <span class="n">tt</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
* original:
AmpliconExperiment (&#34;moving_pic.biom&#34;) with 1967 samples, 7056 features

* filtered:
AmpliconExperiment (&#34;moving_pic.biom&#34;) with 534 samples, 7056 features
</pre></div></div>
</div>
<div class="section" id="we-can-supply-a-list-of-values-instead-of-only-one-value">
<h3>we can supply a list of values instead of only one value<a class="headerlink" href="#we-can-supply-a-list-of-values-instead-of-only-one-value" title="Permalink to this headline">¶</a></h3>
<p>now lets only keep skin and fecal samples</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">print</span><span class="p">(</span><span class="n">movpic</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">[</span><span class="s1">&#39;BODY_HABITAT&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;UBERON:skin&#39; &#39;UBERON:feces&#39; &#39;UBERON:oral cavity&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">yy</span><span class="o">=</span><span class="n">tt</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="s1">&#39;BODY_HABITAT&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;UBERON:skin&#39;</span><span class="p">,</span> <span class="s1">&#39;UBERON:feces&#39;</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="n">yy</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AmpliconExperiment (&#34;moving_pic.biom&#34;) with 399 samples, 7056 features
</pre></div></div>
</div>
</div>
<div class="section" id="we-can-also-reverse-the-filtering-(removing-samples-with-the-supplied-values)">
<h3>we can also reverse the filtering (removing samples with the supplied values)<a class="headerlink" href="#we-can-also-reverse-the-filtering-(removing-samples-with-the-supplied-values)" title="Permalink to this headline">¶</a></h3>
<p>We use the <code class="docutils literal notranslate"><span class="pre">negate=True</span></code> parameter</p>
<p>let’s keep just the non-skin and non-feces samples</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">yy</span><span class="o">=</span><span class="n">tt</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="s1">&#39;BODY_HABITAT&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;UBERON:skin&#39;</span><span class="p">,</span> <span class="s1">&#39;UBERON:feces&#39;</span><span class="p">],</span> <span class="n">negate</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">yy</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AmpliconExperiment (&#34;moving_pic.biom&#34;) with 135 samples, 7056 features
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="filter-low-abundance-features-(filter_abundance)">
<h2>filter low abundance features (<code class="docutils literal notranslate"><span class="pre">filter_abundance</span></code>)<a class="headerlink" href="#filter-low-abundance-features-(filter_abundance)" title="Permalink to this headline">¶</a></h2>
<p>Remove all features (bacteria) with &lt; 10 reads total (summed over all
samples, after normalization).</p>
<p>This is useful for getting rid of non-interesting features. Note that
differently from filtering based of fraction of samples where feature is
present (<code class="docutils literal notranslate"><span class="pre">filter_prevalence</span></code>), this method (<code class="docutils literal notranslate"><span class="pre">filter_abundance</span></code>) will
also keep features present in a small fraction of the samples, but in
high frequency.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">tt</span><span class="o">=</span><span class="n">cfs</span><span class="o">.</span><span class="n">filter_abundance</span><span class="p">(</span><span class="mi">25</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s1">&#39;* original:</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n\n</span><span class="s1">* filtered:</span><span class="se">\n</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">cfs</span><span class="p">,</span> <span class="n">tt</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2018-03-04 12:38:44 INFO After filtering, 766 remaining
* original:
AmpliconExperiment (&#34;chronic-fatigue-syndrome.biom&#34;) with 87 samples, 2129 features

* filtered:
AmpliconExperiment (&#34;chronic-fatigue-syndrome.biom&#34;) with 87 samples, 766 features
</pre></div></div>
</div>
<div class="section" id="Keeping-the-low-abundance-bacteria-instead">
<h3>Keeping the low abundance bacteria instead<a class="headerlink" href="#Keeping-the-low-abundance-bacteria-instead" title="Permalink to this headline">¶</a></h3>
<p>By default, the function removes the low abundance feature. This can be
reversed (i.e. keep low abundance features) by using the <code class="docutils literal notranslate"><span class="pre">negate=True</span></code>
parameter)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">tt</span><span class="o">=</span><span class="n">cfs</span><span class="o">.</span><span class="n">filter_abundance</span><span class="p">(</span><span class="mi">25</span><span class="p">,</span> <span class="n">negate</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s1">&#39;* original:</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n\n</span><span class="s1">* filtered:</span><span class="se">\n</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">cfs</span><span class="p">,</span><span class="n">tt</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2018-03-04 12:38:45 INFO After filtering, 1363 remaining
* original:
AmpliconExperiment (&#34;chronic-fatigue-syndrome.biom&#34;) with 87 samples, 2129 features

* filtered:
AmpliconExperiment (&#34;chronic-fatigue-syndrome.biom&#34;) with 87 samples, 1363 features
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="filter-non-common-bacteria-(filter_prevalence)">
<h2>filter non-common bacteria (<code class="docutils literal notranslate"><span class="pre">filter_prevalence</span></code>)<a class="headerlink" href="#filter-non-common-bacteria-(filter_prevalence)" title="Permalink to this headline">¶</a></h2>
<p>Remove bacteria based on fraction of the samples where this bacteria is
present.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># remove bacteria present in less than half of the samples</span>
<span class="n">tt</span><span class="o">=</span><span class="n">cfs</span><span class="o">.</span><span class="n">filter_prevalence</span><span class="p">(</span><span class="mf">0.5</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s1">&#39;* original:</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n\n</span><span class="s1">* filtered:</span><span class="se">\n</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">cfs</span><span class="p">,</span> <span class="n">tt</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2018-03-04 12:38:46 INFO After filtering, 128 remaining
* original:
AmpliconExperiment (&#34;chronic-fatigue-syndrome.biom&#34;) with 87 samples, 2129 features

* filtered:
AmpliconExperiment (&#34;chronic-fatigue-syndrome.biom&#34;) with 87 samples, 128 features
</pre></div></div>
</div>
</div>
<div class="section" id="Filter-bacteria-based-on-the-mean-frequency-over-all-samples-(filter_mean)">
<h2>Filter bacteria based on the mean frequency over all samples (<code class="docutils literal notranslate"><span class="pre">filter_mean</span></code>)<a class="headerlink" href="#Filter-bacteria-based-on-the-mean-frequency-over-all-samples-(filter_mean)" title="Permalink to this headline">¶</a></h2>
<p>Remove bacteria which have a mean (over all samples) lower than the
desired threshold.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># keep only high frequency bacteria (mean over all samples &gt; 1%)</span>
<span class="n">tt</span><span class="o">=</span><span class="n">cfs</span><span class="o">.</span><span class="n">filter_mean</span><span class="p">(</span><span class="mf">0.01</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s1">&#39;* original:</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n\n</span><span class="s1">* filtered:</span><span class="se">\n</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">cfs</span><span class="p">,</span> <span class="n">tt</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2018-03-04 12:38:47 INFO After filtering, 19 remaining
* original:
AmpliconExperiment (&#34;chronic-fatigue-syndrome.biom&#34;) with 87 samples, 2129 features

* filtered:
AmpliconExperiment (&#34;chronic-fatigue-syndrome.biom&#34;) with 87 samples, 19 features
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2016-2018, calour development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.7.0.<br/>
    </p>
  </div>
</footer>
  </body>
</html>